torch/nn/functional.py

```diff
@@ -3729,6 +3729,50 @@ def triplet_margin_loss(anchor, positive, negative, margin=1.0, p=2, eps=1e-6, s
                                      swap, reduction_enum)
 
 
+def triplet_margin_with_distance_loss(anchor, positive, negative, distance_function=None, is_similarity_function=False,
```

Comment:
@albanD and I had a chance to talk about this more, and we'd like to suggest two changes (for now):

1) making the args after "negative" kwarg-only
2) removing the "is_similarity_function" kwarg

Here's our thinking on not adding "is_similarity_function" at this time:

- We can add it later with no backwards compatibility concern
- We think it's easy for people to convert "similarity functions" to distances, so adding the kwarg now doesn't seem like it would greatly reduce code complexity. For example, `lambda x, y: 1 - cosine_similarity(x, y)` seems comparable to writing `distance_function=cosine_similarity, is_similarity_function=True`. 
- We think asking the user to specify a "distance function" but then let them describe it as a "similarity function" is confusing

What are your thoughts, @ethch18?

Reply:
@mruberry thanks for the feedback!

> making the args after "negative" kwarg-only

Could you expand on the motivation behind this?

> removing the "is_similarity_function" kwarg

This makes sense given that we're not supporting scripting for now and it improves readability.  One concern, since I assume we'd eventually want this to be scriptable: depending on what scripting support for callables looks like later on, making this function scriptable could be BC-breaking for lambda `distance_function`s.  Do you think that will be an issue?

Reply:
> > making the args after "negative" kwarg-only
> 
> Could you expand on the motivation behind this?

Alban and I just have a preference for the clarity of kwargs. We think lots of unnamed arguments is hard to read and if we were designing PyTorch today we'd make better use of kwargs. Kwargs are also nice because they make it easier to change the function signature later.

> > removing the "is_similarity_function" kwarg
> 
> This makes sense given that we're not supporting scripting for now and it improves readability. One concern, since I assume we'd eventually want this to be scriptable: depending on what scripting support for callables looks like later on, making this function scriptable could be BC-breaking for lambda `distance_function`s. Do you think that will be an issue?

I appreciate your thoughtfulness on this but I don't think it will be an issue. If we enable scripting this function then we can say something like, "to script, you must..." and we won't be breaking any existing users.

Reply:
@ethch18 For the kwarg-only, it refers to the new python 3 feature of forcing some args to be passed by kwargs and not by position (see [here](https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html#keyword-only-arguments) for an example).

Reply:
Makes sense.  To make sure I'm on the same page as you re: what the kwargs change would look like:
* change only the function signature, not the docstring
* change both the module and functional signatures
* specify default args via, e.g., `kwargs.get('arg_name', default_val)`

Does this sound right?

Reply:
Yes for the first two points,

> specify default args via, e.g., kwargs.get('arg_name', default_val)

No, this feature is just a 2 character change in the definition that should become:
```python
def triplet_margin_with_distance_loss(anchor, positive, negative, *, distance_function=None, is_similarity_function=False,
                                       margin=1.0, swap=False, reduction="mean"):

Reply:
@albanD whoops, didn't see your comment -- cool feature!  Will add that